staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -222,5 +229,49 @@ var availableAllocators = []struct {
 		) (Allocator, error) {
 			return experimental.NewAllocator(ctx, features, allocateState, classLister, slices, celCache)
 		},
+		nodeMatches: experimental.NodeMatches,
 	},
 }
+
+// NodeMatches determines whether a given Kubernetes node matches the specified criteria.
+// It calls one of the available implementations(stable, incubating, experimental) based
+// on the provided DRA features.
+func NodeMatches(features Features, node *v1.Node, nodeNameToMatch string, allNodesMatch bool, nodeSelector *v1.NodeSelector) (bool, error) {
+	for _, api := range availableAPIs {
+		if api.supportedFeatures.Set().IsSuperset(features.Set()) {
+			return api.nodeMatches(node, nodeNameToMatch, allNodesMatch, nodeSelector)
+		}
+	}
+
+	return false, fmt.Errorf("internal error: no NodeMatches API available for feature set %v", features)
+}
+
+// IsDeviceAllocated checks if a device is allocated, considering both fully allocated devices
+// and partially consumed devices when consumable capacity is enabled.
+func IsDeviceAllocated(deviceID DeviceID, allocatedState *AllocatedState) bool {
+	// Check if device is fully allocated (traditional case)
+	if allocatedState.AllocatedDevices.Has(deviceID) {
+		return true
+	}
+
+	// Check if device is partially consumed via shared allocations (consumable capacity case)
+	// We need to check if any shared device ID corresponds to our device
```

Comment:
```suggestion
	// Check if device is partially consumed via shared allocations (consumable capacity case).
	// We need to check if any shared device ID corresponds to our device.
```

My usual nit :grin: - lack of punctuation. I also prefer to have it elsewhere because one never knows when another (partial) sentence may have to be added.

But this is minor and I am not calling it out elsewhere.